language-plutus-core/src/Language/PlutusCore.hs

```diff
@@ -187,15 +202,21 @@ parseTypecheck
 parseTypecheck gas = typecheckPipeline gas <=< parseScoped
 
 -- | Typecheck a program.
-typecheckPipeline :: (AsNormalizationError e TyName Name a, AsRenameError e a, AsTypeError e a, MonadError e m, MonadQuote m) => Natural -> Program TyName Name a -> m (NormalizedType TyNameWithKind ())
-typecheckPipeline gas p = do
-    checkProgram p
-    typecheckProgram (TypeCheckCfg gas $ TypeConfig False mempty) =<< annotateProgram p
+typecheckPipeline
+    :: (AsNormalizationError e TyName Name a,
+        AsRenameError e a,
+        AsTypeError e a,
+        MonadError e m,
+        MonadQuote m)
+    => Natural
+    -> Program TyName Name a
+    -> m (NormalizedType TyNameWithKind ())
+typecheckPipeline gas = typecheckProgram (TypeCheckCfg gas $ TypeConfig False mempty) <=< annotateProgram <=< through checkProgram
 
 formatDoc :: (AsParseError e AlexPosn, MonadError e m) => BSL.ByteString -> m (Doc a)
 formatDoc bs = runQuoteT $ prettyPlcDef <$> parseProgram bs
 
-format :: (AsParseError e AlexPosn, MonadError e m) => PrettyConfigPlc -> BSL.ByteString -> m T.Text
+format :: (AsParseError e AlexPosn, AsUniqueError e AlexPosn, MonadError e m) => PrettyConfigPlc -> BSL.ByteString -> m T.Text
```

Comment:
We definitely don't want this. Formatting can allow free variables.

Reply:
`UniqueError` isn't just free variables, it's also multiply defined variables and incoherent usages, which I don't think we want to allow.

Reply:
On reflection, I think you're right, this should not run sanity checks. And this means we can include the free variable check by default, which is nice.